### PR TITLE
Make GitHub requests timeout eventually and make Tide throttling configurable.

### PR DESCRIPTION
**Pull request by cjwagner**
The default golang HTTP client doesn't have a timeout set. This makes Prow's GitHub client use a 10 minute timeout. This should never affect a successful request, but should prevent us from blocking forever if GitHub fails to respond. See [this slack conversation](https://kubernetes.slack.com/archives/C7J9RP96G/p1537829113000100).

I changed how Tide throttles a bit because I was seeing us hit the burst limit when fetching PR changes (until they are cached). I also made this configurable via a flag so that we can tweak this without a Prow bump and so that other deployments can use limits that are more sane for their use case.

/cc @BenTheElder @stevekuznetsov @fejta

### CODE REVIEW

File: prow/github/client.go [hunk 1]
```diff
@@ -84,6 +84,10 @@ var (
 
 const (
 	acceptNone = ""
+	// Abort requests that don't return in 10 mins. Longest graphql calls can
+	// take up to 2 minutes. This limit should ensure all successful calls return
+	// but will prevent an indefinite stall if GitHub never responds.
+	maxRequestTime = 10 * time.Minute
```
Comment (BenTheElder) [inline prow/github/client.go]: is 10 minutes really reasonable? that seems excessively long for an HTTP request

Comment (BenTheElder) [reply prow/github/client.go]: OK. 2 to 3 times our already large upper bound seems more reasonable than 5x. Hanging for 10 minutes in the hopes that we might actually get a response before then seems excessive.

Comment (cjwagner) [reply prow/github/client.go]: We could also set a separate limit for v3 and v4 API. v3 should be faster and more reliable so we could set a shorter time limit for the v3 `http.Client` if that is desired, but I don't know what the maximum successful response time is for v3 API requests.

Comment (BenTheElder) [reply prow/github/client.go]: let's get this deployed tomorrow and follow-up with improving it.

File: prow/cmd/tide/main.go [hunk 2]
```diff
@@ -68,6 +71,9 @@ func gatherOptions() options {
 	flag.StringVar(&o.jobConfigPath, "job-config-path", "", "Path to prow job configs.")
 	flag.StringVar(&o.cluster, "cluster", "", "Path to kube.Cluster YAML file. If empty, uses the local cluster.")
 
+	flag.IntVar(&o.syncThrottle, "sync-hourly-tokens", 800, "The number of tokens per hour to be used by the sync controller.")
```
Comment (BenTheElder) [inline prow/cmd/tide/main.go]: minor nit: this is the upper bound right? maybe "The maximum number of tokens per hour .."? between the flag and the description this sort of reads like a target. same for the flag below.

Comment (BenTheElder) [reply prow/cmd/tide/main.go]: these also probably need a note in announcements?

**Comment by stevekuznetsov** [comment]
/lgtm

Agree with Ben on the comments re: flags

File: prow/cmd/tide/main.go (hunk 2 above)
Comment (cjwagner) [reply prow/cmd/tide/main.go]: > these also probably need a note in announcements?

I don't think so? This change is not breaking and only slightly increases the burst size limit for the sync controller, the hourly limits are unchanged. This shouldn't affect any other users of Prow unless flags are specified.

**Pull request merged**